File: leetcode/medium/ReorganizeString.py
from collections import Counter

class Solution:
    def reorganizeString(self, s: str) -> str:
        """
        Rearranges the characters of a string so that no two adjacent characters are the same.

        If a valid rearrangement is not possible, returns an empty string.

        The approach is a greedy one:
        1. Count the frequency of each character in the string.
        2. Check if a solution is possible. A solution is impossible if the count of the
           most frequent character is greater than (N + 1) / 2, where N is the
           length of the string. This is because there wouldn't be enough other
           characters to place between the instances of the most frequent character.
        3. Place the characters into a result array. To ensure no two same characters
           are adjacent, we place them greedily starting with the most frequent ones.
           We fill the even-indexed positions first (0, 2, 4, ...), and then the
           odd-indexed positions (1, 3, 5, ...). This strategy naturally separates
           the most frequent characters as much as possible.

        Time Complexity: O(N), where N is the length of the string. This is because
                         counting characters is O(N) and filling the result is also O(N).
                         The sorting of characters by frequency is O(K log K), where K is
                         the number of unique characters (a constant, max 52), so it
                         doesn't dominate the complexity.
        Space Complexity: O(N) to store the result string.
        """
        # Step 1: Count the frequency of each character.
        counter = Counter(s)
        n = len(s)

        # Step 2: Check if a reorganization is possible.
        # Get the most common character and its count.
        # counter.most_common(1) returns a list like [('a', 5)]
        most_common = counter.most_common(1)
        if most_common:
            _, most_frequent_count = most_common[0]
            if most_frequent_count > (n + 1) // 2:
                return ""

        # Step 3: Greedily place characters into the result array.
        res = [""] * n
        index = 0

        # Iterate through characters sorted by frequency in descending order.
        for char, count in counter.most_common():
            for _ in range(count):
                # Place the character at the current even index.
                res[index] = char
                index += 2
                # If we've gone past the end of the array,
                # wrap around to the first odd index.
                if index >= n:
                    index = 1
        
        # Join the characters in the list to form the final string.
        return "".join(res)

File: leetcode/medium/ReorganizeString.java
class Solution {
    /**
     * Reorganizes the string such that no two adjacent characters are the same.
     * This is possible only if the frequency of the most frequent character 
     * is not greater than (length + 1) / 2.
     *
     * @param s The input string consisting of lowercase English letters.
     * @return A reorganized string, or an empty string if not possible.
     */
    public String reorganizeString(String s) {
        int length = s.length();
        // 1. Count frequencies and find the most frequent character.
        // Array to store frequency of each character ('a' to 'z').
        int[] frequencyMap = new int[26];

        int highestFrequencyCount = 0;
        char highestFrequencyChar = 'a'; // Initialize to a default character

        for (char c : s.toCharArray()) {
            int index = c - 'a';
            frequencyMap[index]++;

            if (frequencyMap[index] > highestFrequencyCount) {
                highestFrequencyCount = frequencyMap[index];
                highestFrequencyChar = c;
            }
        }

        // 2. Check for the impossible case.
        // If the max frequency is greater than the allowed limit (ceil(length / 2)),
        // it's impossible to reorganize, so return "".
        if (highestFrequencyCount > (length + 1) / 2) {
            return "";
        }

        // 3. Place the most frequent character first (Greedy approach).
        char[] result = new char[length];
        
        // Start filling at index 0, then 2, 4, ... (even indices).
        int index = 0; 
        
        // Place the most frequent character into the even positions.
        while (highestFrequencyCount > 0) {
            result[index] = highestFrequencyChar;
            index += 2;
            
            // Wrap around to odd positions (1, 3, 5, ...) if even positions are exhausted.
            if (index >= length) {
                index = 1;
            }
            highestFrequencyCount--;
        }

        // 4. Place the remaining characters.
        // Iterate through all character frequencies.
        for (int i = 0; i < 26; i++) {
            int currentFrequency = frequencyMap[i];
            char currentChar = (char) (i + 'a');

            // The highestFrequencyChar has already been placed, so we skip it 
            // if we didn't track the remaining frequency in the map.
            // A cleaner approach is to use the actual remaining frequency:
            if (currentChar == highestFrequencyChar) {
                // We've already placed the required number of these characters. 
                // Any remaining frequency in the map should be 0 now 
                // (which is true if the loop condition `currentFrequency > 0` is checked 
                // against the *original* frequency after the high-freq char is placed,
                // or simply by skipping it). We skip it here to avoid re-placing.
                continue; 
            }
            
            // Place the remaining characters into the next available slot (which will be odd,
            // or even if index wrapped around)
            while (currentFrequency > 0) {
                result[index] = currentChar;
                index += 2;
                
                // Wrap around to the start of the odd positions if needed.
                if (index >= length) {
                    index = 1;
                }
                currentFrequency--;
            }
        }

        return String.valueOf(result);
    }
}